File: notification/src/main/java/com/ezbuy/notification/client/Impl/AuthClientImpl.java
package com.ezbuy.notification.client.Impl;

import com.ezbuy.framework.client.BaseRestClient;
import com.ezbuy.framework.utils.DataUtil;
import com.fasterxml.jackson.core.type.TypeReference;
import com.ezbuy.notification.client.AuthClient;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.context.annotation.DependsOn;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Slf4j
@Service
@RequiredArgsConstructor
@DependsOn("webClientFactory")
public class AuthClientImpl implements AuthClient {

    @Qualifier(value = "auth")
    private final WebClient auth;
    private final BaseRestClient baseRestClient;

    @Override
    public Mono<List<String>> getAllUserId() {
        return baseRestClient
                .get(auth, "/auth/get-all", null, null, String.class)
                .map(response -> {
                    Optional<String> optionalS = (Optional<String>) response;
                    return DataUtil.parseStringToObject(DataUtil.safeToString(optionalS.get()), new TypeReference<>() {
                    }, new ArrayList<>());
                })
                .onErrorResume(throwable -> {
                    log.error("throwable: ", throwable);
                    return new ArrayList<>();
                });
    }
}

File: notification-service/src/main/kotlin/com/ezbuy/notificationservice/client/impl/AuthClientImpl.kt
package com.ezbuy.notificationservice.client.impl

import com.ezbuy.notificationservice.client.AuthClient
import com.fasterxml.jackson.core.type.TypeReference
import io.hoangtien2k3.reactify.DataUtil
import io.hoangtien2k3.reactify.client.BaseRestClient
import lombok.RequiredArgsConstructor
import lombok.extern.slf4j.Slf4j
import org.reflections.Reflections.log
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.context.annotation.DependsOn
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient
import reactor.core.publisher.Mono
import java.util.*
import kotlin.collections.ArrayList

@Slf4j
@Service
@DependsOn("webClientFactory")
class AuthClientImpl(
    private val baseRestClient: BaseRestClient<String>,
    @Qualifier("auth") private val auth: WebClient
) : AuthClient {

    override fun getAllUserId(): Mono<List<String>> {
        return baseRestClient
            .get(auth, "/auth/get-all", null, null, String::class.java)
            .map { response ->
                val optionalS = response as? Optional<String>
                DataUtil.parseStringToObject(
                    DataUtil.safeToString(optionalS?.get()),
                    object : TypeReference<List<String>>() {},
                    ArrayList()
                )
            }
            .onErrorResume { throwable ->
                log.error("throwable: $throwable")
                Mono.just(ArrayList())
            }
    }
}
